docs/builtin-plugins.rst

```diff
@@ -198,7 +201,99 @@ and start reading. Below is an example python entry point:
     with open(sys.argv[1]) as fp:
         fp.read() # This will be reading the files contents
 
+.. _docker-progress:
+
+Streaming progress reporting from Docker tasks to Girder jobs
+-------------------------------------------------------------
+
+The :py:class:`girder_worker.docker.transforms.girder.ProgressPipe` class can be used to facilitate
+streaming real-time progress reporting from a docker task to its associated Girder job. It uses
+a named pipe to provide a simplistic interface within the container that is usable from any
+runtime environment.
+
+The following example code shows the Girder side task invocation for using ``ProgressPipe``:
+
+.. code-block:: python
+
+    from girder_worker.docker.tasks import docker_run
+    from girder_worker.docker.transforms.girder import ProgressPipe
+
+    docker_run.delay('my_docker_image:latest', container_args=[ProgressPipe()])
+
+The corresponding example code running in the container entrypoint uploads progress events
+at regular intervals, which will automatically reflect in the job progress on the Girder server.
+This code is shown in python, but the idea is the same regardless of language.
+
+.. code-block:: python
+
+    import json
+    import sys
+    import time
+
+    with open(sys.argv[1], 'w') as pipe:
+        for i in range(10):
+            pipe.write(json.dumps({
+                'message': 'Step %d of 10' % i,
+                'total': 10,
+                'current': i + 1
+            }))
+            pipe.flush()
+            time.sleep(1)
+
+The messages written to the pipe must be one per line, and each message must be a JSON Object
+containing optional ``message``, ``current``, and ``total`` values. You **must** call ``flush()``
+on the file handle explicitly for your message to be flushed, since it is a named pipe.
+
+Attaching intermediate / optional artifacts to Girder jobs
+----------------------------------------------------------
+
+It's often useful for debugging/tracing or algorithm analysis to be able to inspect intermediate
+outputs or other artifacts produced during execution of a task, even (perhaps especially) if the
+task fails. These artifacts differ from normal output transforms that upload files to Girder in
+two ways. Firstly, they are optional; if the specified file or directory does not exist, it does
+not cause any errors. This allows docker image authors to choose either at build time or runtime
+whether or not to emit artifacts. Secondly, the emitted files are attached to the job document
```

Comment:
consider: 

"This allows docker image authors to choose either at build time or runtime whether or not to create and upload artifacts. Secondly, the artifact files are attached to the job document..."